### PR TITLE
fix: Client.stream writableNeedDrain

### PR DESCRIPTION
**Pull request by ronag**
Fixes: https://github.com/nodejs/undici/issues/441
Refs: https://github.com/nodejs/node/pull/35348
Refs: https://github.com/nodejs/node/issues/35341

### CODE REVIEW

**Comment by ronag** [comment]
Missing tests.

**Comment by ronag** [comment]
@mcollina: PTAL

File: lib/core/client.js [hunk 1]
```diff
@@ -376,9 +375,37 @@ class Parser extends HTTPParser {
     this.shouldKeepAlive = false
     this.read = 0
     this.request = null
+    this.paused = false
+    this.queue = []
```
Comment (ronag) [inline lib/core/client.js:379]: We can't pause the parser from within a callback so we need to implement a buffer in JS land in order to stop further progress while paused.

**Comment by ronag** [comment]
@indutny @addaleax I'm working around the fact that I can't pause the parser in a callback. Do you have any better suggestions on how to go about that?

**Review by mcollina — CHANGES_REQUESTED** [review]
I don't understand this fix. Why is this needed at all?

(A slightly better implementation would go and replace the kOn* methods when paused to avoid the ifs on an hot path).

**Comment by ronag** [comment]
> I don't understand this fix. Why is this needed at all?

Because currently when pausing we don't actually pause. It's more of a hint.

> A slightly better implementation would go and replace the kOn* methods when paused to avoid the ifs on an hot path

I think this is negligable.

**Comment by ronag** [comment]
I guess it's not entirely necessary... though the current implementation isn't strictly correct. The docs would need to be updated to reflect that returning `false` from `onBody` is only a hint.

**Comment by mcollina** [comment]
Have you benchmarked this?

**Comment by ronag** [comment]
Yes, difference is within margin of error.

**Comment by ronag** [comment]
I've cleaned this up a bit and also added more assertions to make it clearer what this fixes.

File: lib/core/client.js (hunk 1 above)
Comment (mcollina) [reply lib/core/client.js:379]: Can you add a comment so it's clear to future readers?

File: test/client-stream.js [hunk 2]
```diff
@@ -657,3 +657,46 @@ test('stream body destroyed on invalid callback', (t) => {
...
+
+test('stream needDrain', (t) => {
+  t.plan(1)
+
+  const server = createServer((req, res) => {
+    res.end(Buffer.alloc(4096))
+  })
+  t.tearDown(server.close.bind(server))
+
+  server.listen(0, async () => {
```
Comment (mcollina) [inline test/client-stream.js:669]: Don't pass async here, it encourages bad practices about promises.

File: test/client-stream.js [hunk 3]
```diff
@@ -657,3 +657,46 @@ test('stream body destroyed on invalid callback', (t) => {
...
+    }, () => {
+      return dst
+    })
+
+    setTimeout(() => {
+      dst.write = (...args) => {
+        orgWrite.call(dst, ...args)
+      }
+      dst.resume()
+    }, 1e3)
```
Comment (mcollina) [inline test/client-stream.js:696]: is 1s needed?

**Review by mcollina — APPROVED** [review]
LGTM

**Pull request merged**